### PR TITLE
Update indices resolution to be clearer

### PR DESCRIPTION
**Pull request by peternied**
Resolving index patterns into concrete indices had some confusing logic
make it hard to verify its behavior.  Updated this code to be easier
to read and added test cases to confirm resolution patterns.  Also added
unit tests to ensure line/by line coverage was correct.

### Check List
- [X] New functionality includes testing
- [ ] ~New functionality has been documented~
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by cwperks — COMMENTED** [review]
Thank you for the additional tests around the index pattern resolution logic. This is a good addition to making it clearer about how the process works.

File: src/test/java/org/opensearch/security/dlic/dlsfls/FlsIndexingTests.java
Comment (peternied) [reply src/test/java/org/opensearch/security/dlic/dlsfls/FlsIndexingTests.java:51]: We already have many tests for other FLS scenarios, see [FlsTest.java](https://github.com/opensearch-project/security/blob/main/src/test/java/org/opensearch/security/dlic/dlsfls/FlsTest.java) that covers more standard scenarios.  There are also a host of additional tests cases in the [dlsfls](https://github.com/opensearch-project/security/tree/main/src/test/java/org/opensearch/security/dlic/dlsfls) folder.

File: src/test/java/org/opensearch/security/securityconf/impl/v7/IndexPatternTests.java
Comment (peternied) [reply src/test/java/org/opensearch/security/securityconf/impl/v7/IndexPatternTests.java:109]: To me this is a philosophical point, on line 111 in practice the return result should not be altered; however, how can the unit test *prove* that it indeed follow the expected calling pattern setup by the mocks.

This is a case where I would prefer that calls to `resolver.concreteIndexNames(...)` are altered in an unrealistic way to ensure the proper test coverage is obtained.  If someone was to copy/paste this test IMO there would be much higher likelihood they would correctly assert outcomes / verify mock usages.

File: src/test/java/org/opensearch/security/securityconf/impl/v7/IndexPatternTests.java
Comment (peternied) [reply src/test/java/org/opensearch/security/securityconf/impl/v7/IndexPatternTests.java:125]: Same as above, I would prefer to keep the unnaturally aligned as they are.

File: src/test/java/org/opensearch/security/securityconf/impl/v7/IndexPatternTests.java
Comment (peternied) [reply src/test/java/org/opensearch/security/securityconf/impl/v7/IndexPatternTests.java:145]: These calls are slightly unrelated, without doing more footwork in the mock to read the passed in parameters and then return the expected output param.  In my experience the verify on the mock is sufficient to ensure these are aligned and used correctly.

The scenario described for testing for is much better captured in the `FlsIndexingTests.java` as it verifies everything works in practice, when working with several mocks its quickly diminishing returns to align them together.

**Review by cwperks — APPROVED** [review]

**Review by cliu123 — DISMISSED** [review]

**Comment by cliu123** [comment]
The fix looks good to me! But the CI failed. I'll approve when CI passes.

File: src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java
Comment (cliu123) [inline src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java:788]: When ‘resolvedIndices’ is empty, even if ‘appendUnresolved’ is false, unresolved would be still appended with this conditions.

Comment (peternied) [reply src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java:788]: Yes, this behavior is relyed upon in some test cases, so I've make sure to preserve it.

**Review by cliu123 — APPROVED** [review]

File: src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java
Comment (cliu123) [inline src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java:789]: @peternied I've apporved the issue as this is not going to be an immidiate issue blocking this PR. But If there's no need to delay the return to line 791, it'd be safer to return the result at this line as what could happen is that when some indices are appended to `resolvedIndices` between `resolvedIndices.add(unresolved);` and `return resolvedIndices.build();` by accident, then those indices would unexpectedly included in the return result. Please consider changing this as a follow up.

**Pull request merged**
